garak/probes/goat.py

```diff
@@ -0,0 +1,420 @@
+            goal=goal,
+            attack_techniques=attack_techniques_text
+        )
+
+        # AttackerLLM conversation history C_A = [S_A, a_initial_prompt(O)]
+        initial_prompt = GOAT_INITIAL_PROMPT.format(goal=goal)
+        C_A: List[garak.attempt.Turn] = [
+            garak.attempt.Turn("system", garak.attempt.Message(system_prompt)),
+            garak.attempt.Turn("user", garak.attempt.Message(initial_prompt))
+        ]
+
+        # TargetLLM conversation history C_T = []
+        C_T: List[garak.attempt.Turn] = []
+
+        # for k turns in K do
+        for k in range(self.max_iterations):
+            # Generate adversarial prompt/response r_A ← AttackerLLM(C_A, a_name, a_defn, O)
+            attacker_response = self._call_attacker_llm(garak.attempt.Conversation(C_A))
+            if not attacker_response:
+                logging.warning(f"Failed to generate attacker response at turn {k}")
+                continue
+
+            attacker_response_text = self._return_content(attacker_response)
+            # Parse structured output (O, T, S, R)
+            r_A = self._parse_otsr_response(attacker_response_text)
+            if not r_A:
+                logging.warning(f"Failed to parse O-T-S-R from: {attacker_response_text[:200]}")
+                # Fallback: treat entire response as R
+                r_A = {
+                    'Observation': '',
+                    'Thought': '',
+                    'Strategy': '',
+                    'Response': attacker_response_text
+                }
+            if not r_A.get('Response'):
+                logging.warning(f"Failed to generate attack at turn {k}")
+                continue
+
+            # C_A += [r_A] (append complete reasoning)
+            attacker_response_full = self._format_otsr(r_A)
+            C_A.append(garak.attempt.Turn("assistant", garak.attempt.Message(attacker_response_full)))
+
+            # C_T += [r_A['Response']] (append just adversarial prompt)
+            adversarial_prompt = r_A['Response']
+            C_T.append(garak.attempt.Turn("user", garak.attempt.Message(adversarial_prompt)))
+
+            # create attempt with full conversation history
+            target_conversation = garak.attempt.Conversation(C_T)
+            attempt = self._mint_attempt(
+                prompt=target_conversation,
+                seq=goal_idx * self.max_iterations + k,
+                notes={
+                    "goal": goal,
+                    "turn": k+1,
+                    "observation": r_A.get('Observation', ''),
+                    "thought": r_A.get('Thought', ''),
+                    "strategy": r_A.get('Strategy', ''),
+                }
+            )
+
+            # Generate response r_T ← TargetLLM(C_T)
+            r_T: List[garak.attempt.Message] = target_llm.generate(target_conversation)
+
+            if not r_T or not r_T[0]:
+                logging.warning(f"No response from target at turn {k}")
+                continue
```

Comment:
This seems to be modeled after GCG and other iterative single shot attacks, in the case of multi-turn based techniques every call to the target should result in a returned `Attempt` that contains the response from the target.